### ISSUE TITLE
BUG:  An extra blank line is added before and after a message with block quote is edited reported by @huzaifa-99

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12120]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Create a message with block quote (Send the message)
2. Try editing the message after it is sent.

## Expected Result:
There should be no blank lines before/after the block quote message (in edit mode).

## Actual Result:
There are extra blank lines before/after the block quote message (in edit mode).

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.19-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
![Screenshot 2022-10-25 at 2 25 39 AM](https://user-images.githubusercontent.com/43996225/197796213-041ee874-2f55-4140-a4e8-4371c630e881.png)

https://user-images.githubusercontent.com/43996225/197796218-1f81dc9a-18ca-45f8-a7bd-f34ec6f678a2.mp4
https://user-images.githubusercontent.com/43996225/197796261-28430ce3-93ca-4d49-acbf-c46612507ea2.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @huzaifa-99
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666647074971199

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by conorpendergrast** [comment]
This is not a duplicate, is reproducible, is a complete write-up, was created by a QA tester, no Slack conversation about it. I think this is `External`.

**Comment by conorpendergrast** [comment]
Internal: https://www.upwork.com/ab/applicants/1585068624908439552/job-details
External: https://www.upwork.com/jobs/~01c63baf5aa3db8688

**Comment by varshamb** [comment]
### Proposal

We are adding `\n` before and after I think removing it would work. 

https://github.com/Expensify/expensify-common/blob/a71aad37aaba2b43b3a2d2f4be8052b025b265d3/lib/ExpensiMark.js#L256

**Comment by sobitneupane** [comment]
@varshamb Thanks for your proposal.
> We are adding \n before and after while converting html to markdown , I think removing it would work.

Removing it would create an issue. When plain text is sent along the quoted text, line break will be removed.

https://user-images.githubusercontent.com/25876548/197930646-bff2ac0a-afd9-4791-ba30-f2180abc7afc.mov

**Comment by conorpendergrast** [comment]
@sobitneupane [has "sporadic" availability](https://expensify.slack.com/archives/C02NK2DQWUX/p1666597285560119), so will get back to this next week!

**Comment by sobitneupane** [comment]
@0xmiroslav Thanks for your proposal. Your proposal looks promising.

Can you be more descriptive in your proposal. Will you please explain root cause of this issue? And how your proposal solves the issue?

**Comment by sobitneupane** [comment]
@0xmiroslav Your proposal's first step is enough to solve the issue mentioned in issue description. But we must make sure that it does not cause any regression as suggested change will effect all html to markdown translations.
 
Your proposal's second step does not completely solves the issue that you are mentioning. It is retaining the new line when html is converted to markdown but not the opposite way (i.e., markdown to html)

https://user-images.githubusercontent.com/25876548/199223967-87cb1e01-babe-4ce1-9648-3a4b009263f0.mov

I am not sure weather we want to solve the issue that you mentioned in 2nd step of the proposal here or create a separate issue for it? cc: @MonilBhavsar

**Comment by michaelhaxhiu** [comment]
Hey @conorpendergrast just dropping a note as a reminder to [keep the pressure on](https://expensify.slack.com/archives/C01SKUP7QR0/p1667064948124109) to find a contributor and get this one closed out!

cc @MonilBhavsar to weigh in on the latest question to move this forward, and ideally select a proposal + begin PR phase 🙏

**Comment by MonilBhavsar** [comment]
> Your proposal's first step is enough to solve the issue mentioned in issue description. But we must make sure that it does not cause any regression as suggested change will effect all html to markdown translations.

I strongly agree with this. The function is utilized by all rules, so we should evaluate if it doesn't cause any regression on other markdown rules.

@0xmiroslav could you please explain what exactly you're trying to fix in your second proposal.

> And regarding regression test, 1st step also fixes other similar issues along with this GH.

In the video, the blank line still appears no? Is it before the fix?
Regarding regression, I would say in testing steps we check for all markdowns affected by this change.

**Comment by MonilBhavsar** [comment]
Okay got it. I think we should fix that as a part of this issue only as they are almost similar.

**Comment by sobitneupane** [comment]
> Okay got it. I think we should fix that as a part of this issue only as they are almost similar.

@MonilBhavsar In that case, I believe we should update the issue description to include [this case](https://github.com/Expensify/App/issues/12120#issuecomment-1303052492) as well.

https://user-images.githubusercontent.com/25876548/199932697-9b2169be-092b-4c39-8b7d-145c737e95ce.mov

**Comment by MonilBhavsar** [comment]
Ok, something like a blank line is removed/added when block quote is edited?

**Comment by conorpendergrast** [comment]
Looks like this is making progress, let's keep it going to find a proposal that we'll accept!

**Comment by conorpendergrast** [comment]
@sobitneupane bump on getting this proposal making progress 🙏

**Comment by sobitneupane** [comment]
Waiting for proposal.

**Comment by conorpendergrast** [comment]
@sobitneupane @MonilBhavsar So the first part of [this proposal](https://github.com/Expensify/App/issues/12120#issuecomment-1296888030) is not a go? Sorry, having trouble following!

**Comment by sobitneupane** [comment]
Yes, It is not. We are looking for proposal which solves issues for both type of messages (A and B) mentioned [here](https://github.com/Expensify/App/issues/12120#issuecomment-1305590364). The proposal only gives solution for type A message in https://github.com/Expensify/App/issues/12120#issuecomment-1305590364

**Comment by conorpendergrast** [comment]
Thanks so much, I appreciate it! I'll bump up the price to $500 give that it's been more than a week without an accepted solution

**Comment by bernhardoj** [comment]
### **Proposal**
For the first case, we can do the trim as the last step of `htmlToMarkdown` function.

**_expensify-common/ExpensiMark.js_**
```diff
htmlToMarkdown(htmlString) {
        let generatedMarkdown = htmlString;
        const body = /<(body)(?:"[^"]*"|'[^']*'|[^'"><])*>(?:\n|\r\n)?([\s\S]*?)(?:\n|\r\n)?<\/\1>(?![^<]*(<\/pre>|<\/code>))/im;
        const parseBodyTag = generatedMarkdown.match(body);

        // If body tag is found then use the content of body rather than the whole HTML
        if (parseBodyTag) {
            generatedMarkdown = parseBodyTag[2];
        }

        this.htmlToMarkdownRules.forEach((rule) => {
            // Pre-processes input HTML before applying regex
            if (rule.pre) {
                generatedMarkdown = rule.pre(generatedMarkdown);
            }
            generatedMarkdown = generatedMarkdown.replace(rule.regex, rule.replacement);
        });
        generatedMarkdown = generatedMarkdown.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>|<h2>|<\/h2>|<h3>|<\/h3>|<h4>|<\/h4>|<h5>|<\/h5>|<h6>|<\/h6>/gm, '[block]');
-       return Str.htmlDecode(this.replaceBlockWithNewLine(Str.stripHTML(generatedMarkdown)));
+       return Str.htmlDecode(this.replaceBlockWithNewLine(Str.stripHTML(generatedMarkdown))).trim();
    }
```

For the second case, here is what I found.
When we send this **message**
```
Line1
 
 
> Quote text
 
 
Line2
```
the html representation is: `Line1<br /><br /><blockquote>Quote text</blockquote><br /><br />Line2`.
When we press edit, `htmlToMarkdown()` is called which used `htmlToMarkdownRules` to do the conversion. The rule for blockquote will match additional line break before the blockquote which in results remove a line before it. To fix it, we can simply remove the line break match from the regex.
```diff
{
    name: 'quote',
-   regex: /\n?<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
+   regex: /<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
    replacement: (match, g1, g2) => {
        const resultString = g2.trim().split('\n').map(m => `> ${m}`).join('\n');
        return `\n${resultString}\n`;
    },
},
```
We want to match the additional before blockquote because previously, there is this [issue](https://github.com/Expensify/App/issues/9379) where when we send above **message**, the html representation is `Line1<br /><br /><br /><blockquote>Quote text</blockquote><br /><br />Line2`. Notice that there is an extra `<br />` before `<blockquote>`. However, the issue has been fixed by removing the extra `<br />` [here](https://github.com/Expensify/expensify-common/pull/455).

> @0xmiroslav Your proposal's first step is enough to solve the issue mentioned in issue description. But we must make sure that it does not cause any regression as suggested change will effect all html to markdown translations.
> 
> Your proposal's second step does not completely solves the issue that you are mentioning. It is retaining the new line when html is converted to markdown but not the opposite way (i.e., markdown to html)
> 
>  Screen.Recording.2022-11-01.at.17.14.55.mov 
> I am not sure weather we want to solve the issue that you mentioned in 2nd step of the proposal here or create a separate issue for it? cc: @MonilBhavsar

Now, for the last issue where the line is missing before the blockquote on the chat (not in edit mode), I found out that that it only happens on web platform. A blockquote should be on its own line, but the behavior is different between native and web.
Here is what I tested by writing this html `Line1<br /><br /><blockquote>Quote text</blockquote><br /><br />Line2` on Edge.
![image](https://user-images.githubusercontent.com/50919443/202897151-d0ffec95-6a17-4631-a1af-129cfab5950c.png)
You can see that the line before blockquote is missing, while on native the line is shown correctly. The same thing happens on Chrome, Safari, and Firefox. So, the solution I could think of is to add additional `<br />` before blockquote only for web platform.
```
if (Platform.OS === 'web') {
    html = Str.replaceAll(html, /<br \/><blockquote>/gi, '<br /><br /><blockquote>');
}
```
https://github.com/Expensify/App/blob/a3bf7a0f1f3c56e58f97836bf46d09acd010daae/src/pages/home/report/ReportActionItemFragment.js#L101-L124

**Result**

https://user-images.githubusercontent.com/50919443/202898294-1e5c7354-383d-45b8-b5d2-6009041d35f3.mov


https://user-images.githubusercontent.com/50919443/202899101-cbd6b557-1628-45e4-8c48-65e214bdc7d5.mp4

**Comment by bernhardoj** [comment]
I don't agree that it is the issue of `react-native-render-html`. I have provided a screenshot on my proposal showing how it looks on web from a plain .html file.

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/ExpensiMark.js b/ExpensiMark2.js
index 129fca3..8f04da1 100644
--- a/ExpensiMark.js
+++ b/ExpensiMark2.js
@@ -184,11 +184,6 @@ export default class ExpensiMark {
                 regex: /<br\s*[/]?><pre>\s*/gi,
                 replacement: ' <pre>',
             },
-            {
-                name: 'replacebrquote',
-                regex: /<br\s*[/]?><blockquote>\s*/gi,
-                replacement: '<blockquote>',
-            },
         ];
 
         /**
@@ -250,10 +245,16 @@ export default class ExpensiMark {
             },
             {
                 name: 'quote',
-                regex: /\n?<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
-                replacement: (match, g1, g2) => {
-                    const resultString = g2.trim().split('\n').map(m => `> ${m}`).join('\n');
-                    return `\n${resultString}\n`;
+                regex: /(.?)<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\2>(.?)(?![^<]*(<\/pre>|<\/code>))/gi,
+                replacement: (match, g1, g2, g3, g4) => {
+                    let resultString = g3.trim().split('\n').map(m => `> ${m}`).join('\n');
+                    if (g1) {
+                        resultString = `${g1}\n${resultString}`;
+                    }
+                    if (g4) {
+                        resultString = `${resultString}\n${g4}`;
+                    }
+                    return resultString;
                 },
             },
             {
@@ -544,7 +545,7 @@ export default class ExpensiMark {
                     // Make sure we will only modify if we have Text needed to be formatted for quote
                     if (textToFormat !== '') {
                         textToFormat = textToFormat.replace(/\n$/, '');
-                        replacedText += replacement(textToFormat);
+                        replacedText += `${replacement(textToFormat)}\n`;
                         textToFormat = '';
                     }
```

### Details
The first issue is caused by the newlines as first discovered by @varshamb
The fix is to use another two capturing group as helpers to decide when a newline is necessary. This is better than trimming the resulting output because trim only works on leading and trialling spaces (i.e. may fail if there is another text other than the blockquote and the two empty lines)

The second issue is partly caused by a missing newline at `modifyTextForQuote()` function and mostly by the rule `replacebrquote`, and since it's behaviour seems to be unwanted I removed it.

Modified file: https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js

**Comment by conorpendergrast** [comment]
Looks like a lot for @sobitneupane to review here! Thanks all

**Comment by sobitneupane** [comment]
Testing both the above proposals across all platforms.

**Comment by sobitneupane** [comment]
@s77rt Your proposal is adding extraline both above and below the quote while converting markdown to html.
> https://user-images.githubusercontent.com/25876548/202997524-de123a91-e8c9-40bd-a0e7-e93d6480bda3.mov

**Comment by sobitneupane** [comment]
> So, the solution I could think of is to add additional `<br />` before blockquote only for web platform.

@bernhardoj We don't prefer platform specific code? Are you sure it is absolutely necessary?

**Comment by sobitneupane** [comment]
@s77rt Can you please include video result as well (if possible)? And please do include why each change you are proposing is necessary and what you are trying to achieve with the change.

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/ExpensiMark.js b/ExpensiMark2.js
index 129fca3..109bc22 100644
--- a/ExpensiMark.js
+++ b/ExpensiMark2.js
@@ -186,9 +186,14 @@ export default class ExpensiMark {
             },
             {
                 name: 'replacebrquote',
-                regex: /<br\s*[/]?><blockquote>\s*/gi,
+                regex: /(?:<br\s*[/]?>)+<blockquote>\s*/gi,
                 replacement: '<blockquote>',
             },
+            {
+                name: 'replacequotebr',
+                regex: /\s*<\/blockquote>(?:<br\s*[/]?>)+/gi,
+                replacement: '</blockquote>',
+            },
         ];
 
         /**
@@ -250,10 +255,16 @@ export default class ExpensiMark {
             },
             {
                 name: 'quote',
-                regex: /\n?<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\1>(?![^<]*(<\/pre>|<\/code>))/gi,
-                replacement: (match, g1, g2) => {
-                    const resultString = g2.trim().split('\n').map(m => `> ${m}`).join('\n');
-                    return `\n${resultString}\n`;
+                regex: /(.?)<(blockquote|q)(?:"[^"]*"|'[^']*'|[^'">])*>([\s\S]*?)<\/\2>(.?)(?![^<]*(<\/pre>|<\/code>))/gi,
+                replacement: (match, g1, g2, g3, g4) => {
+                    let resultString = g3.trim().split('\n').map(m => `> ${m}`).join('\n');
+                    if (g1) {
+                        resultString = `${g1}\n${resultString}`;
+                    }
+                    if (g4) {
+                        resultString = `${resultString}\n${g4}`;
+                    }
+                    return resultString;
                 },
             },
             {
@@ -544,7 +555,7 @@ export default class ExpensiMark {
                     // Make sure we will only modify if we have Text needed to be formatted for quote
                     if (textToFormat !== '') {
                         textToFormat = textToFormat.replace(/\n$/, '');
-                        replacedText += replacement(textToFormat);
+                        replacedText += `${replacement(textToFormat)}\n`;
                         textToFormat = '';
                     }
```

### Update
Replace all `<br>`s before and after `<blockquote>`.
Replacing only one will cause a problem on editing, for every edit another newline will be removed e.g.: if the initial text has 4 empty lines, it will be 3, and on the edit it will be 2, if you edit it once again it will be 1 and one more time the left new line will be removed.
Removing all the newlines seems to be the way to go, also it's the same here in github.

### Edits:
1. `modifyTextForQuote` function is missing a newline to be added after the `replacedText` inside the loop. I have tested this function independently. The newline is required.
2. Adjusted the `quote` rule regex to match an additional character before and after the quote so we know when an additional newline is required. I have also removed the newline match at first because it's matching a required newline and not a redundant one.
3. Adjusted `replacebrquote` to match multiple `br` tags as stated in the update above
4. Added `replacequotebr`, it does the same work as the rule before but works for the trialling `br` tags



https://user-images.githubusercontent.com/16493223/203018567-80f3b654-0efe-4f4e-a064-4c6dae8c3147.mp4

**Comment by MonilBhavsar** [comment]
> @MonilBhavsar I suggest to handle these 2 issues separately. Reason:

Ah ok, so do you think the second issue affects all markdowns?

**Comment by MonilBhavsar** [comment]
> One thing for sure is, we need to handle it differently on web and native, as it behaves differently. It's not a library issue, but a web behavior. I have tested it by creating an .html file containing this text `Line1<br /><br /><blockquote>Quote text</blockquote><br /><br />Line2` and the line before blockquote is discarded (tested on Chrome, Edge, Firefox, Safari).

@bernhardoj your reasoning looks good. But this issue is also reproducible on native App. So I don't think the platform specific fix would work

**Comment by bernhardoj** [comment]
@MonilBhavsar The line discarded issue on native is only when we editing the message, but it is fine showing the message.

**Comment by iwiznia** [comment]
Context https://expensify.slack.com/archives/C01GTK53T8Q/p1669040394837259

> so long as the edited version is semantically the same as the original (eg, both produce the same HTML), it's not a bug.

So I think we should close this.
